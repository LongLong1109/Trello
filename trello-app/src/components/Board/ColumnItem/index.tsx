import { useCallback } from 'react'
import { Flex, Card, CloseButton } from '@mantine/core'

// components
import { Input, Button } from '@/components/common'

interface ColumnItemProps {
  listName: string
  onAddList: (listName: string) => void
  onIsAddingList: (isAdding: boolean) => void
  onListName: (listName: string) => void
}

const ColumnItem = ({ onAddList, onIsAddingList, listName, onListName }: ColumnItemProps) => {
  const handleAddList = useCallback(() => {
    if (listName.trim()) {
      onAddList(listName)
      onListName('')
      onIsAddingList(false)
    }
  }, [listName, onAddList, onIsAddingList, onListName])

  const handleCloseAddingList = () => {
    onIsAddingList(false)
  }

  return (
    <Card shadow='md' w='270' radius='md'>
      <Flex direction='column' gap='sm'>
        <Input
          variant='primary'
          placeholder='Enter list title...'
          value={listName}
          onChange={({ currentTarget: { value } }) => onListName(value)}
        />
        <Flex direction='row' gap='sm' align='center'>
          <Button onClick={handleAddList} name='Add List' />

          <CloseButton onClick={handleCloseAddingList} />
        </Flex>
      </Flex>
    </Card>
  )
}

export default ColumnItem
