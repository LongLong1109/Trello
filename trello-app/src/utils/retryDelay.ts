export const retryDelay = (attemptIndex: number) => Math.min(1000 * 2 ** attemptIndex, 30000)
