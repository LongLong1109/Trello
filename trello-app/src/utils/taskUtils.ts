//interface
import { ListTask, TaskListItem, Task } from '@/interfaces/Task'
import { TASK_PROPERTY_KEY, TaskPropertyValue } from '@/types/Task'

export const updateTaskProperty = (
  lists: ListTask[],
  listId: number,
  taskId: number,
  key: TASK_PROPERTY_KEY,
  value: TaskPropertyValue,
) => {
  return lists.map((list) => {
    const isTargetList = list.id === listId
    if (isTargetList) {
      return {
        ...list,
        tasks: list.tasks.map((task) => (task.id === taskId ? { ...task, [key]: value } : task)),
      }
    }
    return list
  })
}

export const removeTaskById = (lists: ListTask[], listId: number, taskId: number) => {
  return lists.map((list) => {
    const isTargetList = list.id === listId
    if (isTargetList) {
      return {
        ...list,
        tasks: list.tasks.filter((task) => task.id !== taskId),
      }
    }
    return list
  })
}

export const updateTaskTitle = (lists: ListTask[], selectedTask: TaskListItem, title: string) => {
  return lists.map((list) =>
    list.id === selectedTask.listId
      ? {
          ...list,
          tasks: list.tasks.map((task) =>
            task.id === selectedTask.id ? { ...task, title } : task,
          ),
        }
      : list,
  )
}

export const moveTask = (
  taskToMove: Task,
  updatedLists: ListTask[],
  targetListId: number,
  targetIndex: number,
) => {
  return updatedLists.map((list) => {
    if (list.id === targetListId) {
      const updatedTasks = [...list.tasks]
      updatedTasks.splice(targetIndex, 0, taskToMove)
      return { ...list, tasks: updatedTasks }
    }
    return list
  })
}
