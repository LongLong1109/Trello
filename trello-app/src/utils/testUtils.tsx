import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { MantineProvider } from '@mantine/core'
const AllBrowserRouter = ({ children }: { children: React.ReactNode }) => {
  return <MantineProvider>{children}</MantineProvider>
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllBrowserRouter, ...options })

export * from '@testing-library/react'
export * from '@testing-library/user-event'
export { customRender as render }
