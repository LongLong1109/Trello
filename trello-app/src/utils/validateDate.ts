export const isDateOverdue = (date: Date) => {
  const currentDate = new Date()
  currentDate.setHours(0, 0, 0, 0)
  date.setHours(0, 0, 0, 0)

  const timeDifference = date.getTime() - currentDate.getTime()
  return timeDifference < 0
}

export const convertDateToString = (date: Date) => {
  const validDate = date ? new Date(date) : new Date()
  return validDate.toLocaleDateString()
}

export const formatDate = (date: Date): string => {
  if (!date || isNaN(date.getTime())) {
    return '---'
  }

  const options: Intl.DateTimeFormatOptions = { month: 'short', day: 'numeric' }
  return new Intl.DateTimeFormat('en-US', options).format(date)
}
