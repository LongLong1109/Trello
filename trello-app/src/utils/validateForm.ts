import { MESSAGES } from '@/constants/message'
import { REGEX_EMAIL, PASSWORD_MIN_LENGTH } from '@/constants/validateFields'

export const validateFirstName = (value: string) => {
  return value === '' ? MESSAGES.firstName : null
}

export const validateLastName = (value: string) => {
  return value === '' ? MESSAGES.lastName : null
}

export const validateEmail = (value: string) => {
  return REGEX_EMAIL.test(value) ? null : MESSAGES.email
}

export const validatePassword = (value: string) => {
  if (value === '') {
    return MESSAGES.passwordRequired
  }
  return value.length >= PASSWORD_MIN_LENGTH ? null : MESSAGES.password
}
