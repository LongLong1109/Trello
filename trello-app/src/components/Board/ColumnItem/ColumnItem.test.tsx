// components
import ColumnItem from '../ColumnItem';

// utils
import { render, screen, fireEvent } from '@/utils/testUtils';

describe('ColumnItem component', () => {
  const mockOnAddList = jest.fn();
  const mockOnIsAddingList = jest.fn();
  const mockOnListName = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const renderColumnItem = (isAddingList: boolean, listName: string) => {
    return render(
      <ColumnItem
        isAddingList={isAddingList}
        listName={listName}
        onAddList={mockOnAddList}
        onIsAddingList={mockOnIsAddingList}
        onListName={mockOnListName}
      />,
    );
  };

  it('should be matches snapshot', () => {
    const { asFragment } = renderColumnItem(false, 'Title');
    expect(asFragment()).toMatchSnapshot();
  });

  it('should renders the add button when isAddingList is false', () => {
    renderColumnItem(false, 'Title');

    expect(screen.getByText('Add a list')).toBeInTheDocument();
    expect(screen.queryByPlaceholderText('Enter list title...')).toBeNull();
  });

  it('should call onAddList with the correct arguments and reset states', () => {
    renderColumnItem(true, 'New List');
    fireEvent.click(screen.getByText('Add List'));
    expect(mockOnAddList).toHaveBeenCalledWith('New List');
    expect(mockOnListName).toHaveBeenCalledWith('');
    expect(mockOnIsAddingList).toHaveBeenCalledWith(false);
  });

  it('should not call onAddList if the input is empty', () => {
    renderColumnItem(true, '');
    fireEvent.click(screen.getByText('Add List'));
    expect(mockOnAddList).not.toHaveBeenCalled();
  });
});
