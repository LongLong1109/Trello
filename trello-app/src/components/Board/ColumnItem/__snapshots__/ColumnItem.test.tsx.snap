// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnItem component should be matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <button
    class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
    data-variant="light-white"
    data-with-left-section="true"
    justify="flex-start"
    style="--button-justify: flex-start; --button-radius: var(--mantine-radius-lg);"
    type="button"
  >
    <span
      class="m_80f1301b mantine-Button-inner"
    >
      <span
        class="m_a74036a mantine-Button-section"
        data-position="left"
      >
        <svg
          class="tabler-icon tabler-icon-plus "
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 5l0 14"
          />
          <path
            d="M5 12l14 0"
          />
        </svg>
      </span>
      <span
        class="m_811560b9 mantine-Button-label"
      >
        Add a list
      </span>
    </span>
  </button>
</DocumentFragment>
`;
